Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift

```diff
@@ -259,43 +313,69 @@ public class Mqtt5Client {
                 let continuationCore = ContinuationCore(continuation: continuation)
                 callbackOptions.completion_callback = unsubscribeCompletionCallback
                 callbackOptions.completion_user_data = continuationCore.passRetained()
-                let result = aws_mqtt5_client_unsubscribe(rawValue, unsubscribePacketPointer, &callbackOptions)
-                guard result == AWS_OP_SUCCESS else {
-                    continuationCore.release()
-                    return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                self.rwlock.read {
+                    // validate the client in case close() is called.
+                    guard let rawValue = self.rawValue else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
+                    let result = aws_mqtt5_client_unsubscribe(rawValue, unsubscribePacketPointer, &callbackOptions)
+                    guard result == AWS_OP_SUCCESS else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
                 }
             }
         }
     }
 
+    /// Discard all operations and cleanup the client. It is MANDATORY function to call to release the client.
     public func close() {
-        self.callbackCore.close()
-        aws_mqtt5_client_release(rawValue)
-        rawValue = nil
+        self.rwlock.write {
+            self.callbackFlag = false
+            aws_mqtt5_client_release(rawValue)
+            self.rawValue = nil
+        }
+    }
+
+    /////////////////////////////////////////////////////////
+    // helper functions for self retained reference
+    private func callbackUserData() -> UnsafeMutableRawPointer {
+        return Unmanaged<Mqtt5Client>.passRetained(self).toOpaque()
+    }
+
+    private func release() {
```

Comment:
This is only ever called in one place if we fail to create a new native client. Let's just execute this code there instead of having it as a separate function.